### ISSUE TITLE
Graph is already complete when API requests lands from previously-completed stages (?)

### ISSUE THREAD

**Issue by jan-g** [fnproject/flow#111]
Under load, I'm seeing a few 400 responses from the "Vista demo" load test.

It looks like the following is happening:

- stage 56 is launched successfully (but doesn't finish yet)
- stage 56 completes
- other stages finish
- termination stage 0 is run
- the POST response is recorded by GIN.

I must confess that I don't understand this sequence of events; how is it possible that an API request for stage 56 does not complete until after the stage has finished?
[400-error.log](https://github.com/fnproject/flow/files/1461307/400-error.log)

... it must be another stage. Need to work out what's going on here.

**Comment by jan-g** [comment]
It looks to me like the stage that caused the attempt to create a thenComplete node errored out (no logs though). In the example I'm looking at currently That happened about 10s (!) before the terminated graph was rematerialised by an attempt to handle that POST request.

Could it be that our requests are stuck in a queue/goroutine/something between GIN accepting the request, a socket being disconnected due to container exit and then the request floating around inside the flow service until it finally gets addressed? (!)

**Comment by jan-g** [comment]
To verify this:

1. To make it easier to track we need the call ID logging at debug.
2. Ideally we want to log at the moment the HTTP request is accepted, not just when we're finished and about to return a response to it.

**Comment by hhexo** [comment]
Here's my hypothesis for the sequence of events that triggers this:

- Flow initiates running stage X and calls Fn to run it
- Stage X's container initiates a thenCompose API call
- Flow receives the API call and the http connection is established
- Flow starts to process the API call
- The API call should be synchronous, but the network connection transiently goes down
- Stage X sees the connection go down and fails with a PlatformException
- While stage X's API call is still running, Flow detects stage X has failed and updates the graph
- The graph does not yet have the node created by the thenCompose API call
- Flow marks the graph as complete because there are no more nodes
- The API call is finally processed and fails because the graph is already complete

In other words, I believe it's a race condition when transient network errors occur and the container has a different world view than Flow.

To test it, I'm going to add logging to the vista-like test so every api call is wrapped in a try/catch and we can check what the stage thinks has happened to the API call.

**Comment by zootalures** [comment]
All things being equal - if flows perceives that a stage has failed before the container completes then this is currently "correct behaviour"

**Comment by hhexo** [comment]
Yes, the moment there's a network drop on the http call then the world view of Flow and its client start to diverge, so weird stuff may start to happen.

The user's function sees a "failure while adding a stage to the flow" which then may have to be recovered from. But the recovery logic cannot know if the stage was added or not anyway - this is essentially a "failure to call a remote non-idempotent api", you don't know what happened to the remote side.
So if a user sees an error of this kind, it is irrelevant whether the API call succeeded or failed because they still can't do anything about it.

Of course, this still looks bad for our reliability story. In @jan-g 's test, 241 calls out of 63000 failed with this problem (then causing those flows to fail) - that's about 99.6% reliability. We need more nines. :)

**Comment by jan-g** [comment]
Precise details aside, that rest response time (ie, round trip from our rest handler - ideally from inside GIN)  should also go into prometheus. Some of the broken ones hovered at 5-6s (!) and I'd like to work out why.


-- Excuse typos. Phone screens don't make great ketbroads. 

-------- Original message --------

**Comment by jan-g** [comment]
This is down to Fn returning 504 (timeout) whilst the function continues to run in its hot container. See https://github.com/fnproject/fn/issues/497
